cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,139 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+    if (currentSpec == null) {
+      LOG.info("Application {} not found for upgrade.", appId);
+      return new ApplicationUpdateDetail(appId, new NotFoundException(appId));
+    }
+    ArtifactId currentArtifact = currentSpec.getArtifactId();
+    NamespaceId currentArtifactNamespace =
+      ArtifactScope.SYSTEM.equals(currentArtifact.getScope()) ? NamespaceId.SYSTEM : appId.getParent();
+
+    // Get the artifact with latest version for upgrade.
+    List<ArtifactSummary> availableArtifacts =
+      artifactRepository.getArtifactSummaries(currentArtifactNamespace, currentArtifact.getName(), 1,
+                                              ArtifactSortOrder.DESC);
+    if (availableArtifacts.isEmpty()) {
+      // This should not be possible as at least the current used artifact should be there.
+      String error = String.format("No artifacts found for artifact id %s in namespace %s.", currentArtifact.getName(),
+                                   currentArtifactNamespace);
+      return new ApplicationUpdateDetail(appId, "Upgrade failed.", error);
+    }
+    // The latest version should be first (and only) value in the result.
+    ArtifactSummary candidateArtifact = availableArtifacts.get(0);
+    ArtifactVersion candidateArtifactVersion = new ArtifactVersion(candidateArtifact.getVersion());
+
+    // Current artifact should not have higher version than candidate artifact.
+    if (currentArtifact.getVersion().compareTo(candidateArtifactVersion) > 0) {
+      String error = String.format(
+        "Requested artifact version %s is older than current artifact version %s.", candidateArtifact.getVersion(),
+        currentArtifact.getVersion());
+      return new ApplicationUpdateDetail(appId, "Upgrade failed." , error);
+    }
+
+    ArtifactId newArtifactId =
+        new ArtifactId(currentArtifact.getName(), candidateArtifactVersion, currentArtifact.getScope());
+
+    Id.Artifact newArtifact = Id.Artifact.fromEntityId(Artifacts.toProtoArtifactId(appId.getParent(), newArtifactId));
+    ArtifactDetail newArtifactDetail = artifactRepository.getArtifact(newArtifact);
+    List<ApplicationConfigUpdateAction> upgradeActions = Arrays.asList(ApplicationConfigUpdateAction.UPGRADE_ARTIFACT);
+
+
+    return updateApplicationInternal(appId, appId.getParent(), appId.getApplication(), null,
+                                    currentSpec.getConfiguration(), programTerminator, newArtifactDetail,
+                                    upgradeActions, ownerAdmin.getOwner(appId), /*updateSchedules=*/false, "Upgrade");
+  }
+
+  /**
+   * Updates an application config by applying given update actions. The app should know how to apply these actions
+   * to its config.
+   */
+  private ApplicationUpdateDetail updateApplicationInternal(ApplicationId applicationId,
+                                                             NamespaceId namespaceId,
+                                                             @Nullable String appName,
```

Comment:
this can't be null right?

Reply:
also, shouldn't need to pass in namespace, app, or version is ApplicationId is passed in.